### DISCUSSION TITLE
Incorrect redirect_uri

### DISCUSSION THREAD

**Discussion by slominskir** [keycloak/keycloak#25684] [Q&A] [answered]
I am attempting to upgrade to the latest Keycloak (v23.0.3) from an older working v20.0.2 instance.  Simply using the same redirect_uri setting for a client doesn't work anymore.    The client is running on Wildly v26.1.3 with a [patched](https://github.com/slominskir/wildfly-elytron/releases/tag/v1.19.1.Patch1) Elytron.   Is there some trick to configuring validRedirectUri in Keycloak or is this combination of Elytron and Keycloak a no-go?

The error in the Wildfly log when I attempt to login:
```
2023-12-18 16:23:10,874 ERROR [org.wildfly.security.http.oidc] (default task-6) status from server: 400
2023-12-18 16:23:10,874 ERROR [org.wildfly.security.http.oidc] (default task-6)    {"error":"invalid_grant","error_description":"Incorrect redirect_uri"}
```

I've tried adding a bunch of valid redirect URIs and even tried a single asterisk, which apparently doesn't mean anything goes:

- `https://proxy.example.com/app/*`
- `https://backendserver.example.com/app/*`
- `*`
- `+`
- `https://proxy.example.com/*`

I guess I'll stick to the old version for now, but eventually I need to patch and move forward.  What gives?   Is 23.0.3 busted?  Maybe the existence of URL parameters in the return_uri is no longer tolerated by Keycloak?

The login returns back to the app with "Forbidden" displayed (Wildfly/Elytron generated HTML) and the URL has the original page params intact plus additional Keycloak params:
```
https://proxy.example.com/app?appParam1=...&state=...&session_state=...&iss=https%3A%2F%2Fproxy.example.com%2Fauth%2Frealms%2Fmyrealm&code=...
```

My understanding is the recent best practice is for page URL parameters to be bundled up into the `state` param.  However, Elytron in Wildfly v26.1.3 doesn't do that and instead by default strips URL parameters.   Keycloak used to be OK with URL parameters, but it appears maybe not so anymore.   Not clear if latest Elytron (Wildfly 30.x) supports maintaining page URL parameters in the `state` param.  Returning users to the page they were on before clicking login with URL parameters intact seems like a reasonable behavior to maintain.

**Comment by slominskir** [comment]
Related:
- https://github.com/keycloak/keycloak/discussions/9278
- https://github.com/keycloak/keycloak/discussions/25002
- https://github.com/keycloak/keycloak/discussions/24687
- https://github.com/keycloak/keycloak/issues/24851
- https://github.com/keycloak/keycloak/issues/25637

There is a lot of activity related to `redirect-uri` lately, and the best I can tell it's now broken in that it no longer allows the wildcard (*) to mean allow URL parameters (multi-page apps) or URL fragments (single page apps).

**Comment by rmartinc** [comment]
@slominskir Yes, there have been several security issues so now the comparison is more strict for sure. But for understanding this better, can you please paste the exact redirectUri that is in the browser if possible? (Just copy the redirect query param as it's in the browser).

**Reply by slominskir** [reply]
Are you suggesting the contents of the URL parameters matter?    The format of the URL as mentioned above is: `https://proxy.example.com/app?appParam1=...&state=...`.   

The contents often do contain URL encoded (escaped) characters as the apps I maintain support pages that can be viewed both read-only or as an authenticated user.  To support this a declarative security path setting for all pages in web.xml would not work and instead a single dedicated declaratively protected Servlet can be requested by clicking a login link and that Servlet will then use a URL parameter to determine which page in the app to return to.  See:

[Working Docker example (with Keycloak 20.0.5)](https://github.com/JeffersonLab/calendar)

If you run the demo containers above with `docker compose up` the URLs you'll see are:
- Start on app page showing month of Dec here: `https://localhost:8443/calendar/view-month?calendar=1&month=12&year=2023`
- Login button click while on page for Dec. results in browser redirect to `https://localhost:8443/calendar/sso?returnUrl=https%3A%2F%2Flocalhost%3A8443%2Fcalendar%2Fview-month%3Fcalendar%3D1%26month%3D12%26year%3D2023`
- Wildfly/Elytron security auth constraint on that Servlet results in redirect to: `http://localhost:8081/realms/test-realm/protocol/openid-connect/auth?response_type=code&client_id=calendar&redirect_uri=https%3A%2F%2Flocalhost%3A8443%2Fcalendar%2Fsso%3FreturnUrl%3Dhttps%253A%252F%252Flocalhost%253A8443%252Fcalendar%252Fview-month%253Fcalendar%253D1%2526month%253D12%2526year%253D2023&state=7f4f43c3-d87b-454d-a71d-7b4e3d4412a5&scope=openid`
- Typing demo username and password in Keycloak form and clicking submit results in post to: `http://localhost:8081/realms/test-realm/login-actions/authenticate?session_code=1Q9YSKz32rUMwQ9AQ1NxCKxIKnPfd5BK7Mdwu2KsyQ4&execution=048d5c12-c666-4dd6-9563-5179fade88d9&client_id=calendar&tab_id=Ij-9asfSM80`
- Which redirects back to the app at: `https://localhost:8443/calendar/sso?returnUrl=https%3A%2F%2Flocalhost%3A8443%2Fcalendar%2Fview-month%3Fcalendar%3D1%26month%3D12%26year%3D2023&state=7f4f43c3-d87b-454d-a71d-7b4e3d4412a5&session_state=b23d9fd8-920b-4b62-a477-091ee58b643c&code=6ca31b03-1091-4494-881e-0002ddd2f124.b23d9fd8-920b-4b62-a477-091ee58b643c.ab992e78-6d60-40d7-8014-dc6d3ade4116`
- Which Wilfly/Elytron redirects back to: `https://localhost:8443/calendar/sso?returnUrl=https%3A%2F%2Flocalhost%3A8443%2Fcalendar%2Fview-month%3Fcalendar%3D1%26month%3D12%26year%3D2023`
- And finally the SSORedirect Servlet redirects the browser back to the original page the user was on (URL params intact plus now authenticated): `https://localhost:8443/calendar/view-month?calendar=1&month=12&year=2023`

Ideally you could  simply bump Keycloak to version 23.0.3 in [deps.yml](https://github.com/JeffersonLab/calendar/blob/81c6ac707b3aa487475faef2495811e3c0ec1563/deps.yml#L16) to see the fail state.  This turns out to be tricky since latest Keycloak container [no longer contains curl or a package manager](https://www.keycloak.org/server/containers) and therefore requires a custom extended Dockerfile to do even basic stuff like check if the container started (so you can then configure it programmatically).   Seems like latest version of Keycloak breaks fairly basic stuff like returning users to the page they were on before auth with params intact and allowing Keycloak container to determine if the server is fully started so that configs can be applied all in the name of hardening and being more secure.

**Reply by slominskir** [reply]
I've added a branch named `keycloak23` that has the custom Dockerfile needed to support Keycloak 23 such that the fail case can be observed:

https://github.com/JeffersonLab/calendar/compare/main...keycloak23

```
git clone -b keycloak23 https://github.com/JeffersonLab/calendar.git
cd calendar
docker compose up
```

**Comment by rmartinc** [comment]
@slominskir I have done a little test with 23.0.3 and using the url you set in point 3: `http://localhost:8080/realms/test-realm/protocol/openid-connect/auth?response_type=code&client_id=calendar&redirect_uri=https%3A%2F%2Flocalhost%3A8443%2Fcalendar%2Fsso%3FreturnUrl%3Dhttps%253A%252F%252Flocalhost%253A8443%252Fcalendar%252Fview-month%253Fcalendar%253D1%2526month%253D12%2526year%253D2023&state=7f4f43c3-d87b-454d-a71d-7b4e3d4412a5&scope=openid` I can perfom the login. I have created a `calendar` client with valid redirect URI as `https://localhost:8443/calendar/sso/*`. Using `*` it works in the same way. After doing the login (username and password), the browser is redirected to `https://localhost:8443/calendar/sso?returnUrl=https%3A%2F%2Flocalhost%3A8443%2Fcalendar%2Fview-month%3Fcalendar%3D1%26month%3D12%26year%3D2023&state=7f4f43c3-d87b-454d-a71d-7b4e3d4412a5&session_state=65ebaf07-8582-4fd9-a56d-742ee23674f4&iss=http%3A%2F%2Flocalhost%3A8080%2Frealms%2Ftest-realm&code=b1ffa282-e41a-492f-81c8-0a82df1c7bda.65ebaf07-8582-4fd9-a56d-742ee23674f4.a0d3eeeb-45e8-4a33-b508-c84f7eeaad69` which doesn't exists in my env. But the login is done successfully:

```
2023-12-20 08:52:13,347 TRACE [org.keycloak.events] (executor-thread-14) type=LOGIN, realmId=90e50af5-afe1-43fb-a79e-2cc616c0f06c, clientId=calendar, userId=bebcda52-2de8-4f12-9a5c-2addbe376ee7, ipAddress=127.0.0.1, auth_method=openid-connect, auth_type=code, response_type=code, redirect_uri=https://localhost:8443/calendar/sso?returnUrl=https%3A%2F%2Flocalhost%3A8443%2Fcalendar%2Fview-month%3Fcalendar%3D1%26month%3D12%26year%3D2023, consent=no_consent_required, code_id=50ac8706-d485-4b17-9279-b6534e58a0c5, username=ricky, response_mode=query, authSessionParentId=50ac8706-d485-4b17-9279-b6534e58a0c5, authSessionTabId=6PZJDb5ENEU, requestUri=http://localhost:8080/realms/test-realm/login-actions/authenticate?session_code=DtyH3MD6Wi56wMq4JGF_k44kQ37DszDENnnuZgQJUiU&execution=58456275-a8ea-48a1-bce7-950f0222b564&client_id=calendar&tab_id=6PZJDb5ENEU, cookies=[AUTH_SESSION_ID_LEGACY=50ac8706-d485-4b17-9279-b6534e58a0c5, KC_RESTART=eyJhbGciOiJIUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICIxNmFlODc3YS02OTU3LTQ1YzktODU1YS0zNWMzM2Q4MTVhY2QifQ.eyJjaWQiOiJjYWxlbmRhciIsInB0eSI6Im9wZW5pZC1jb25uZWN0IiwicnVyaSI6Imh0dHBzOi8vbG9jYWxob3N0Ojg0NDMvY2FsZW5kYXIvc3NvP3JldHVyblVybD1odHRwcyUzQSUyRiUyRmxvY2FsaG9zdCUzQTg0NDMlMkZjYWxlbmRhciUyRnZpZXctbW9udGglM0ZjYWxlbmRhciUzRDElMjZtb250aCUzRDEyJTI2eWVhciUzRDIwMjMiLCJhY3QiOiJBVVRIRU5USUNBVEUiLCJub3RlcyI6eyJzY29wZSI6Im9wZW5pZCIsImlzcyI6Imh0dHA6Ly9sb2NhbGhvc3Q6ODA4MC9yZWFsbXMvdGVzdC1yZWFsbSIsInJlc3BvbnNlX3R5cGUiOiJjb2RlIiwicmVkaXJlY3RfdXJpIjoiaHR0cHM6Ly9sb2NhbGhvc3Q6ODQ0My9jYWxlbmRhci9zc28_cmV0dXJuVXJsPWh0dHBzJTNBJTJGJTJGbG9jYWxob3N0JTNBODQ0MyUyRmNhbGVuZGFyJTJGdmlldy1tb250aCUzRmNhbGVuZGFyJTNEMSUyNm1vbnRoJTNEMTIlMjZ5ZWFyJTNEMjAyMyIsInN0YXRlIjoiN2Y0ZjQzYzMtZDg3Yi00NTRkLWE3MWQtN2I0ZTNkNDQxMmE1In19.bc3g0Ep1rSDH9t__AsLEHPReLUrApq3cVYyxgfPCM6A, AUTH_SESSION_ID=50ac8706-d485-4b17-9279-b6534e58a0c5]
```

Are you sure the openidc URL is that one? The error is displayed at that point, right?

**Reply by slominskir** [reply]
Keycloak still redirects back to the app at the URL above even when login is unsuccessful.  Why do you think it was successful (maybe you see a URL parameter or something that suggests the user was authenticated in the app)?   If the user logs into keycloak with the correct password it's possible that the Keycloak session (keycloak server log) says successful even if app client authentication fails due to redirect_uri.  I think you may need to test with a setup that includes a client.

**Reply by rmartinc** [reply]
@slominskir The error of an invalid redirect URI stops the login at that point and shows a page saying the login was unsuccessful because of the redirect_uri. If you pass that step it means that the redirect URI was OK and the login is performed. The error is the following in case of an invalid redirect_uri:

![Screenshot from 2023-12-20 15-31-53](https://github.com/keycloak/keycloak/assets/26901232/f0a1ac62-a229-410c-8f07-41bb7777ab49)

But where is the error for you? In the subsequent redirects?

In my case after the login it tries to go to `https://localhost:8443/calendar/sso?returnUrl=https%3A%2F%2Flocalhost%3A8443%2Fcalendar%2Fview-month%3Fcalendar%3D1%26month%3D12%26year%3D2023&state=7f4f43c3-d87b-454d-a71d-7b4e3d4412a5&session_state=65ebaf07-8582-4fd9-a56d-742ee23674f4&iss=http%3A%2F%2Flocalhost%3A8080%2Frealms%2Ftest-realm&code=b1ffa282-e41a-492f-81c8-0a82df1c7bda.65ebaf07-8582-4fd9-a56d-742ee23674f4.a0d3eeeb-45e8-4a33-b508-c84f7eeaad69` which is your original redirect_uri plus the state, code, iss and the rest of the parameters for the code to token.

**Reply by slominskir** [reply]
The error in the Wildfly log file as mentioned above is:
```
status from server: 400
{"error":"invalid_grant","error_description":"Incorrect redirect_uri"}
```
and Wildfly/Elytron simply returns the Forbidden HTML to the user:
![image](https://github.com/keycloak/keycloak/assets/2642119/567d0611-a09a-4f32-9998-5ceb903adb2f)


So it sounds like the client credentials flow may be where things fall apart (Elytron).

**Reply by rmartinc** [reply]
Then try adding the log `--log-level=INFO,org.keycloak.protocol.oidc.endpoints:TRACE`. That error is returned because the `redirectUri` in the code is different to the one in the parameter. I'm trying to see this debug [line](https://github.com/keycloak/keycloak/blob/23.0.3/services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java#L401). Maybe it's the new `iss` parameter, you can also try to remove it.

**Reply by slominskir** [reply]
```
keycloak  | 2023-12-20 10:40:59,891 TRACE [org.keycloak.protocol.oidc.endpoints.AuthorizationEndpoint] (executor-thread-4) Processing @GET request
keycloak  | 2023-12-20 10:40:59,892 DEBUG [org.keycloak.protocol.oidc.endpoints.AuthorizationEndpointChecker] (executor-thread-4) PKCE non-supporting Client
keycloak  | 2023-12-20 10:40:59,903 WARN  [org.keycloak.events] (executor-thread-4) type=CODE_TO_TOKEN_ERROR, realmId=398cac3b-be13-44ed-8933-2b97d5628cf3, clientId=calendar, userId=c6444a7d-14cd-4033-b023-da3b31d58dde, ipAddress=172.23.0.3, error=invalid_code, grant_type=authorization_code, code_id=b7c820c4-f956-40e4-b365-bff6509e53d0, client_auth_method=client-secret
keycloak  | 2023-12-20 10:40:59,903 TRACE [org.keycloak.protocol.oidc.endpoints.TokenEndpoint] (executor-thread-4) Parameter 'redirect_uri' did not match originally saved redirect URI used in initial OIDC request. Saved redirectUri: https://localhost:8443/calendar/sso?returnUrl=https%3A%2F%2Flocalhost%3A8443%2Fcalendar%2F, redirectUri parameter: https://localhost:8443/calendar/sso?returnUrl=https%3A%2F%2Flocalhost%3A8443%2Fcalendar%2F&iss=http%3A%2F%2Flocalhost%3A8081%2Frealms%2Ftest-realm
calendar  | 10:40:59,904 ERROR [org.wildfly.security.http.oidc] (default task-2) failed to turn code into token
calendar  | 10:40:59,904 ERROR [org.wildfly.security.http.oidc] (default task-2) status from server: 400
calendar  | 10:40:59,904 ERROR [org.wildfly.security.http.oidc] (default task-2)    {"error":"invalid_grant","error_description":"Incorrect redirect_uri"}
```

**Reply by rmartinc** [reply, marked as answer]
Yep! It's the iss, if you see the two redirect_uri are just different because of the iss param. Go to your client and in the `Advanced` tab set `Exclude Issuer From Authentication Response` to ON.

This problem is not related to teh changes in the redirect_uri validation but with the new iss parameter introduced by the spec. Your client does not remove it from the URI.

**Reply by slominskir** [reply]
Wow, thanks so much for figuring this out.  It works now after using the option above.

**Reply by SergioCea** [reply]
Man, I LOVE YOU @rmartinc

**Reply by imdmahajankanika** [reply]
Hello, during the logout from realm application, I am getting 

`Invalid parameter: redirect_uri`

```
2024-03-06 16:50:27,774 WARN  [org.keycloak.events] (executor-thread-9) type="LOGOUT_ERROR", realmId="Cerberus", clientId="null", userId="null", ipAddress="172.16.12.250", error="invalid_request"
2024-03-06 16:50:27,774 WARN  [org.keycloak.protocol.oidc.endpoints.LogoutEndpoint] (executor-thread-9) Parameter 'redirect_uri' no longer supported. Please use 'post_logout_redirect_uri' with 'id_token_hint' for this endpoint. Alternatively you can enable backwards compatibility option 'legacy-logout-redirect-uri' of oidc login protocol in the server configuration.
```

**Reply by slominskir** [reply]
Just a quick note for those attempting to script this configuration setting:

```
/opt/keycloak/bin/kcadm.sh update clients/${CLIENT_ID} -r ${REALM} -s 'attributes."exclude.issuer.from.auth.response"=true'
```

**Reply by mainProgram** [reply]
U THE BEST @rmartinc

**Reply by mhauer-eb** [reply]
That was a great help!
